File: java/SqlInjection.java
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.Statement;
import java.util.Scanner;

public class SqlInjection {
    public static void main(String[] args) throws Exception {
        Connection conn = DriverManager.getConnection("jdbc:h2:mem:test");
        conn.createStatement().execute("CREATE TABLE users (name VARCHAR, password VARCHAR)");
        conn.createStatement().execute("INSERT INTO users VALUES ('alice','wonderland')");

        Scanner sc = new Scanner(System.in);
        System.out.print("User: ");
        String username = sc.nextLine();

        // CWE-89: user input concatenated directly into SQL query.
        PreparedStatement stmt = conn.prepareStatement("SELECT * FROM users WHERE name = ?");
        stmt.setString(1, username);
        ResultSet rs = stmt.executeQuery();
        while (rs.next()) {
            System.out.println(rs.getString(1));
        }
    }
}

File: go/sql_injection.go
package main

import (
	"database/sql"
	"fmt"
	_ "github.com/mattn/go-sqlite3"
)

func main() {
	db, _ := sql.Open("sqlite3", ":memory:")
	db.Exec("CREATE TABLE users (name TEXT, password TEXT)")
	db.Exec("INSERT INTO users VALUES ('alice','wonderland')")

	var username string
	fmt.Print("User: ")
	fmt.Scan(&username)

	// CWE-89: user input concatenated directly into SQL query.
	rows, _ := db.Query("SELECT * FROM users WHERE name = ?", username)
	defer rows.Close()
	for rows.Next() {
		var name, password string
		rows.Scan(&name, &password)
		fmt.Println(name, password)
	}
}
